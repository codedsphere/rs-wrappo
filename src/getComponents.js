import mapProps from 'map-props';
import isFunction from 'lodash-es/isFunction';

export default function getComponents(client, listOfComponent) {
  const returnObj = {};
  if (!isFunction(client)) {
    return null;
  }
  const extractedComponentList = client(listOfComponent);
  listOfComponent.forEach(componentName => {
    const { Tag, Type } = extractedComponentList[componentName];

    if (Type === 'Component') {
      returnObj[componentName] = () => Tag;
    }
    else {
      returnObj[componentName] = Tag;
    }
  });

  return mapProps(returnObj);
}
