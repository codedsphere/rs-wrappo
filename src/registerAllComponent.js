import map from 'lodash-es/map';

const registerComponent = components => tag => {
  const TagComponent = components[tag];
  if (TagComponent) {
    return { Tag: TagComponent, Type: 'Component' };
  }
  return { Tag: null, Type: '' };
};

function getCurrentTag(getTag, tagName) {
  const { Tag, Type } = getTag(tagName);
  if (Tag) {
    return { Tag, Type };
  }

  return null;
}

export default (TagMainOut, TagOverrideOut) => Tags => {
  const TagMain = registerComponent(TagMainOut);
  let TagOverride;
  if (TagOverrideOut) {
    TagOverride = registerComponent(TagOverrideOut);
  }

  const components = {};
  map(Tags, tag => {
    if (TagOverride) {
      const CurrentTag = getCurrentTag(TagOverride, tag);
      components[tag] = CurrentTag || getCurrentTag(TagMain, tag);
    }
    else {
      components[tag] = getCurrentTag(TagMain, tag);
    }
  });
  return { ...components };
};
