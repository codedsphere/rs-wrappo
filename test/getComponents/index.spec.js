import React, { Component } from 'react';
import { mount } from 'enzyme';
import getComponents from 'getComponents';
import registerAllComponent from 'registerAllComponent';
import main from '../__mocks__/client/main';
import override from '../__mocks__/client/override';

describe('getComponents', () => {
  it('should have place only empty value', () => {
    const ExamplePage = getComponents({}, ['assets']);
    expect(ExamplePage).toBe(null);
  });
});

describe('getComponents', () => {
  it('should have place only main value', () => {
    const MainComponentRendered = registerAllComponent(main);

    class Example extends Component {
      render() {
        const { main: Main } = this.props.assets;
        return <Main />;
      }
    }

    const ExamplePage = getComponents(MainComponentRendered, ['assets'])(Example);
    expect(mount(<ExamplePage />)).toMatchSnapshot();
    expect(mount(<ExamplePage />).contains(<div>Test Main</div>)).toBe(true);
  });
});

describe('getComponents', () => {
  it('should have place only override value', () => {
    const MainComponentRendered = registerAllComponent(main, override);

    class Example extends Component {
      render() {
        const { main: Main } = this.props.assets;
        return <Main />;
      }
    }

    const ExamplePage = getComponents(MainComponentRendered, ['assets'])(Example);
    expect(mount(<ExamplePage />)).toMatchSnapshot();
    expect(mount(<ExamplePage />).contains(<div>Test Override</div>)).toBe(true);
  });
});
