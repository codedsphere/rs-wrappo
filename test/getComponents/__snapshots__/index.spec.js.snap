// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponents should have place only main value 1`] = `
<MapProps(Example)>
  <Example
    assets={
      Object {
        "main": [Function],
      }
    }
  >
    <TestPageMain>
      <div>
        Test Main
      </div>
    </TestPageMain>
  </Example>
</MapProps(Example)>
`;

exports[`getComponents should have place only override value 1`] = `
<MapProps(Example)>
  <Example
    assets={
      Object {
        "main": [Function],
      }
    }
  >
    <TestPageMain>
      <div>
        Test Override
      </div>
    </TestPageMain>
  </Example>
</MapProps(Example)>
`;
